### ISSUE TITLE
Token Morpho ypsID Key E not supported (USB device idVendor=1a6f, idProduct=0104)

### ISSUE THREAD

**Issue by achelouah** [OpenSC/OpenSC#1090]
I'm using ccid-1.4.3 patched with [ccid-morpho-v7-2.patch](https://ludovicrousseau.blogspot.fr/2017/02/mostly-ccid-driver-for-some-morpho.html) and opensc git master (commit 4ea2828).
```
**$ pcsc_scan**
PC/SC device scanner
V 1.4.22 (c) 2001-2011, Ludovic Rousseau <ludovic.rousseau@free.fr>
Compiled with PC/SC lite version: 1.8.11
Using reader plug'n play mechanism
Scanning present readers...
0: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00

Wed Jul  5 12:46:37 2017
Reader 0: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00
  Card state: Card inserted, 
  ATR: 3B 7D 18 00 02 80 57 59 50 53 49 44 30 34 83 7F 90 00

ATR: 3B 7D 18 00 02 80 57 59 50 53 49 44 30 34 83 7F 90 00
+ TS = 3B --> Direct Convention
+ T0 = 7D, Y(1): 0111, K: 13 (historical bytes)
  TA(1) = 18 --> Fi=372, Di=12, 31 cycles/ETU
    129032 bits/s at 4 MHz, fMax for Fi = 5 MHz => 161290 bits/s
  TB(1) = 00 --> VPP is not electrically connected
  TC(1) = 02 --> Extra guard time: 2
+ Historical bytes: 80 57 59 50 53 49 44 30 34 83 7F 90 00
  Category indicator byte: 80 (compact TLV data object)
    Tag: 5, len: 7 (card issuer's data)
      Card issuer data: 59 50 53 49 44 30 34
    Tag: 8, len: 3 (status indicator)
      LCS (life card cycle): 7F (Proprietary)
      SW: 9000 (Normal processing.)

Possibly identified card (using /home/chelouah/.cache/smartcard_list.txt):
3B 7D 18 00 02 80 57 59 50 53 49 44 30 34 83 7F 90 00
        Morpho e-CPF YpsID S2-11/11 (PKI)
        http://safeweb.com.br

**$ ./opensc-tool -l**
0x7fa27db6a700 13:01:58.186 [opensc-tool] ctx.c:792:sc_context_create: ===================================
0x7fa27db6a700 13:01:58.186 [opensc-tool] ctx.c:793:sc_context_create: opensc version: 0.17.0
0x7fa27db6a700 13:01:58.186 [opensc-tool] reader-pcsc.c:815:pcsc_init: PC/SC options: connect_exclusive=0 disconnect_action=1 transaction_end_action=0 reconnect_action=0 enable_pinpad=1 enable_pace=1
0x7fa27db6a700 13:01:58.186 [opensc-tool] reader-pcsc.c:1283:pcsc_detect_readers: called
0x7fa27db6a700 13:01:58.186 [opensc-tool] reader-pcsc.c:1302:pcsc_detect_readers: Probing PC/SC readers
0x7fa27db6a700 13:01:58.186 [opensc-tool] reader-pcsc.c:1330:pcsc_detect_readers: Establish PC/SC context
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:1242:pcsc_add_reader: Adding new PC/SC reader 'MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00'
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:319:refresh_attributes: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00 check
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:347:refresh_attributes: current  state: 0x00000022
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:348:refresh_attributes: previous state: 0x00000000
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:403:refresh_attributes: card present, changed
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:1408:pcsc_detect_readers: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00:SCardConnect(SHARED): 0x00000000
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:1063:detect_reader_features: called
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:1065:detect_reader_features: Requesting reader features ... 
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:1086:detect_reader_features: Reader feature 0a found
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:1086:detect_reader_features: Reader feature 12 found
0x7fa27db6a700 13:01:58.855 [opensc-tool] reader-pcsc.c:1153:detect_reader_features: Reader does not have a display.
0x7fa27db6a700 13:01:58.856 [opensc-tool] reader-pcsc.c:1004:part10_detect_max_data: get dwMaxAPDUDataSize property returned -1
0x7fa27db6a700 13:01:58.856 [opensc-tool] reader-pcsc.c:1205:detect_reader_features: Assuming that the reader supports transceiving short length APDUs only
0x7fa27db6a700 13:01:58.856 [opensc-tool] reader-pcsc.c:1043:part10_get_vendor_product: id_vendor=ffffffff id_product=ffffffff
0x7fa27db6a700 13:01:58.858 [opensc-tool] reader-pcsc.c:1423:pcsc_detect_readers: returning with: 0 (Success)
# Detected readers (pcsc)
Nr.  Card  Features  Name
0x7fa27db6a700 13:01:58.858 [opensc-tool] sc.c:275:sc_detect_card_presence: called
0x7fa27db6a700 13:01:58.858 [opensc-tool] reader-pcsc.c:411:pcsc_detect_card_presence: called
0x7fa27db6a700 13:01:58.858 [opensc-tool] reader-pcsc.c:319:refresh_attributes: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00 check
0x7fa27db6a700 13:01:58.858 [opensc-tool] reader-pcsc.c:339:refresh_attributes: returning with: 0 (Success)
0x7fa27db6a700 13:01:58.858 [opensc-tool] reader-pcsc.c:416:pcsc_detect_card_presence: returning with: 1
0x7fa27db6a700 13:01:58.858 [opensc-tool] sc.c:280:sc_detect_card_presence: returning with: 1
0    Yes             MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00
0x7fa27db6a700 13:01:58.858 [opensc-tool] ctx.c:870:sc_release_context: called
0x7fa27db6a700 13:01:58.858 [opensc-tool] reader-pcsc.c:896:pcsc_finish: called

**$ ./opensc-tool --reader 0 --name**
0x7f6b40096700 13:05:47.595 [opensc-tool] card.c:200:sc_connect_card: called
0x7f6b40096700 13:05:47.595 [opensc-tool] card-entersafe.c:138:entersafe_match_card: called
0x7f6b40096700 13:05:47.595 [opensc-tool] card-rutoken.c:103:rutoken_match_card: called
0x7f6b40096700 13:05:47.883 [opensc-tool] card-mcrd.c:296:mcrd_match_card: SELECT AID: 6A86
0x7f6b40096700 13:05:47.945 [opensc-tool] muscle.c:276:msc_select_applet: returning with: -1200 (Card command failed)
0x7f6b40096700 13:05:47.945 [opensc-tool] card-piv.c:2919:piv_match_card: called
0x7f6b40096700 13:05:47.945 [opensc-tool] card-piv.c:761:piv_find_aid: called
0x7f6b40096700 13:05:47.945 [opensc-tool] card-piv.c:3365:piv_card_reader_lock_obtained: called
0x7f6b40096700 13:05:47.967 [opensc-tool] card-piv.c:3365:piv_card_reader_lock_obtained: called
0x7f6b40096700 13:05:47.987 [opensc-tool] card-cac.c:1506:cac_match_card: called
0x7f6b40096700 13:05:47.987 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:47.987 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 1 16 db 0  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:47.987 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.010 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.010 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 0  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.010 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.031 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.031 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 1  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.031 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.053 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.053 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 2  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.053 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.079 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.079 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 3  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.079 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.102 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.102 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 4  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.102 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.122 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.122 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 5  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.122 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.142 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.142 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 6  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.142 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.164 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.164 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 7  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.164 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.185 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.185 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 8  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.185 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.203 [opensc-tool] card-cac.c:1004:cac_select_file_by_type: called
0x7f6b40096700 13:05:48.203 [opensc-tool] card-cac.c:1017:cac_select_file_by_type: path->aid=a0 0 0 0 79 1 9  len=7, path->value = 0 0 0 0 len=0 path->type=1 (1)
0x7f6b40096700 13:05:48.203 [opensc-tool] card-cac.c:1019:cac_select_file_by_type: file_out=(nil) index=0 count=0
0x7f6b40096700 13:05:48.224 [opensc-tool] card-itacns.c:162:itacns_match_card: Matching 3b against atr[0] == 3b
0x7f6b40096700 13:05:48.251 [opensc-tool] card-gids.c:567:gids_match_card: called
0x7f6b40096700 13:05:48.251 [opensc-tool] card-gids.c:280:gids_select_aid: called
0x7f6b40096700 13:05:48.274 [opensc-tool] card-gids.c:298:gids_select_aid: returning with: -1205 **(Incorrect parameters in APDU)**
0x7f6b40096700 13:05:48.338 [opensc-tool] card-coolkey.c:2241:coolkey_match_card: called
0x7f6b40096700 13:05:48.338 [opensc-tool] card-coolkey.c:934:coolkey_apdu_io: called
**Unsupported card**
0x7f6b40096700 13:05:48.894 [opensc-tool] ctx.c:870:sc_release_context: called
```
Il tried to add
```
       card_atr 3B:7D:18:00:02:80:57:59:50:53:49:44:30:34:83:7F:90:00 {
                type = 25005;
                driver = "iasecc";
                name = "Morpho ypsID Key E IAS/ECC";
                # secure_messaging = local_morpho_YpsID_S3;
        }
```
in **opensc.conf** with no success. I always get

> Failed to connect to card: Incorrect parameters in APDU

(I also tried with **type = 25003**)

**Comment by frankmorgner** [comment]
@viktorTarasov do you know this type of IAS/ECC?

**Comment by achelouah** [comment]
Il can provide more informations as far as concern the token if you want

**Comment by dengert** [comment]
It looks like the message "Failed to connect to card: Incorrect parameters in APDU" maybe misleading.  It comes from the card-gids.c driver trying to determine if you have a gids card. Then card-coolkey.c driver is called  and it can not support your card. 

A full opensc-debug.log `debug = 7;` would be more helpful, as I don't see any card-iasecc.c log entries. 
and include the opensc.conf file too.

**Comment by achelouah** [comment]
With `debug = 7` and `debug_file = /tmp/opensc-debug.log` in `opensc.conf`

```
[root@localhost ~]# opensc-tool -l -n
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    Yes             MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00
Using reader with a card: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00
Failed to connect to card: Incorrect parameters in APDU
```
```
[root@localhost bin]# cat /tmp/opensc-debug.log 
0x7f093197b700 17:45:28.498 [opensc-tool] ctx.c:792:sc_context_create: ===================================
0x7f093197b700 17:45:28.498 [opensc-tool] ctx.c:793:sc_context_create: opensc version: 0.17.0
0x7f093197b700 17:45:28.498 [opensc-tool] reader-pcsc.c:815:pcsc_init: PC/SC options: connect_exclusive=0 disconnect_action=1 transaction_end_action=0 reconnect_action=0 enable_pinpad=1 enable_pace=1
0x7f093197b700 17:45:28.498 [opensc-tool] reader-pcsc.c:1283:pcsc_detect_readers: called
0x7f093197b700 17:45:28.499 [opensc-tool] reader-pcsc.c:1302:pcsc_detect_readers: Probing PC/SC readers
0x7f093197b700 17:45:28.499 [opensc-tool] reader-pcsc.c:1330:pcsc_detect_readers: Establish PC/SC context
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1242:pcsc_add_reader: Adding new PC/SC reader 'MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00'
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:319:refresh_attributes: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00 check
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:347:refresh_attributes: current  state: 0x00000022
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:348:refresh_attributes: previous state: 0x00000000
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:403:refresh_attributes: card present, changed
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1408:pcsc_detect_readers: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00:SCardConnect(SHARED): 0x00000000
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1063:detect_reader_features: called
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1065:detect_reader_features: Requesting reader features ... 
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1086:detect_reader_features: Reader feature 0a found
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1086:detect_reader_features: Reader feature 12 found
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1153:detect_reader_features: Reader does not have a display.
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1004:part10_detect_max_data: get dwMaxAPDUDataSize property returned -1
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1205:detect_reader_features: Assuming that the reader supports transceiving short length APDUs only
0x7f093197b700 17:45:29.167 [opensc-tool] reader-pcsc.c:1043:part10_get_vendor_product: id_vendor=ffffffff id_product=ffffffff
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:1423:pcsc_detect_readers: returning with: 0 (Success)
0x7f093197b700 17:45:29.169 [opensc-tool] sc.c:275:sc_detect_card_presence: called
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:411:pcsc_detect_card_presence: called
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:319:refresh_attributes: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00 check
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:339:refresh_attributes: returning with: 0 (Success)
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:416:pcsc_detect_card_presence: returning with: 1
0x7f093197b700 17:45:29.169 [opensc-tool] sc.c:280:sc_detect_card_presence: returning with: 1
0x7f093197b700 17:45:29.169 [opensc-tool] sc.c:275:sc_detect_card_presence: called
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:411:pcsc_detect_card_presence: called
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:319:refresh_attributes: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00 check
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:339:refresh_attributes: returning with: 0 (Success)
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:416:pcsc_detect_card_presence: returning with: 1
0x7f093197b700 17:45:29.169 [opensc-tool] sc.c:280:sc_detect_card_presence: returning with: 1
0x7f093197b700 17:45:29.169 [opensc-tool] sc.c:275:sc_detect_card_presence: called
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:411:pcsc_detect_card_presence: called
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:319:refresh_attributes: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00 check
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:339:refresh_attributes: returning with: 0 (Success)
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:416:pcsc_detect_card_presence: returning with: 1
0x7f093197b700 17:45:29.169 [opensc-tool] sc.c:280:sc_detect_card_presence: returning with: 1
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:200:sc_connect_card: called
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:533:pcsc_connect: called
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:319:refresh_attributes: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00 check
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:339:refresh_attributes: returning with: 0 (Success)
0x7f093197b700 17:45:29.169 [opensc-tool] reader-pcsc.c:565:pcsc_connect: Initial protocol: T=0
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1057:match_atr_table: ATR     : 3b:7d:18:00:02:80:57:59:50:53:49:44:30:34:83:7f:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:00:70:0A:90:00:8B
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1057:match_atr_table: ATR     : 3b:7d:18:00:02:80:57:59:50:53:49:44:30:34:83:7f:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:7F:96:00:00:00:31:B9:64:40:70:14:10:73:94:01:80:82:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:7F:96:00:00:00:31:B8:64:40:70:14:10:73:94:01:80:82:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:7D:18:00:02:80:57:59:50:53:49:44:30:34:83:7F:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:225:sc_connect_card: matching configured ATRs
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:234:sc_connect_card: trying driver 'authentic'
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1057:match_atr_table: ATR     : 3b:7d:18:00:02:80:57:59:50:53:49:44:30:34:83:7f:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:00:70:0A:90:00:8B
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:234:sc_connect_card: trying driver 'iasecc'
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1057:match_atr_table: ATR     : 3b:7d:18:00:02:80:57:59:50:53:49:44:30:34:83:7f:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:7F:96:00:00:00:31:B9:64:40:70:14:10:73:94:01:80:82:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:7F:96:00:00:00:31:B8:64:40:70:14:10:73:94:01:80:82:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:7D:18:00:02:80:57:59:50:53:49:44:30:34:83:7F:90:00
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:239:sc_connect_card: matched driver 'IAS-ECC'
0x7f093197b700 17:45:29.169 [opensc-tool] card-iasecc.c:345:iasecc_match_card: iasecc_match_card(3B7D180002805759505349443034837F 9000) called
0x7f093197b700 17:45:29.169 [opensc-tool] card.c:1057:match_atr_table: ATR     : 3b:7d:18:00:02:80:57:59:50:53:49:44:30:34:83:7f:90:00
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:7F:96:00:00:00:31:B8:64:40:70:14:10:73:94:01:80:82:90:00
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:08:00:07:90:00:FE
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:7D:13:00:00:4D:44:57:2D:49:41:53:2D:43:41:52:44:32
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:7F:18:00:00:00:31:B8:64:50:23:EC:C1:73:94:01:80:82:90:00
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:DF:96:00:80:31:FE:45:00:31:B8:64:04:1F:EC:C1:73:94:01:80:82:90:00:EC
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:DF:18:FF:81:91:FE:1F:C3:00:31:B8:64:0C:01:EC:C1:73:94:01:80:82:90:00:B3
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:DC:18:FF:81:91:FE:1F:C3:80:73:C8:21:13:66:02:04:03:55:00:02:34
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1068:match_atr_table: ATR try : 3B:DC:18:FF:81:91:FE:1F:C3:80:73:C8:21:13:66:01:0B:03:52:00:05:38
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1071:match_atr_table: ignored - wrong length
0x7f093197b700 17:45:29.170 [opensc-tool] card-iasecc.c:348:iasecc_match_card: card not matched
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:257:sc_connect_card: driver 'IAS-ECC' match_card() failed: Success (will continue anyway)
0x7f093197b700 17:45:29.170 [opensc-tool] card-iasecc.c:632:iasecc_init: called
0x7f093197b700 17:45:29.170 [opensc-tool] card-iasecc.c:523:iasecc_init_sagem: called
0x7f093197b700 17:45:29.170 [opensc-tool] card-iasecc.c:365:iasecc_parse_ef_atr: called
0x7f093197b700 17:45:29.170 [opensc-tool] ef-atr.c:147:sc_parse_ef_atr: called
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:748:sc_select_file: called; type=2, path=3f002f01
0x7f093197b700 17:45:29.170 [opensc-tool] card-iasecc.c:867:iasecc_select_file: called
0x7f093197b700 17:45:29.170 [opensc-tool] card-iasecc.c:874:iasecc_select_file: iasecc_select_file(card:0x1187740) path.len 4; path.type 2; aid_len 0
0x7f093197b700 17:45:29.170 [opensc-tool] card-iasecc.c:875:iasecc_select_file: iasecc_select_file() path:3f002f01
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:1244:sc_print_cache: card cache invalid
0x7f093197b700 17:45:29.170 [opensc-tool] card-iasecc.c:879:iasecc_select_file: EF.ATR(aid:'')
0x7f093197b700 17:45:29.170 [opensc-tool] card-iasecc.c:246:iasecc_select_mf: called
0x7f093197b700 17:45:29.170 [opensc-tool] apdu.c:554:sc_transmit_apdu: called
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:407:sc_lock: called
0x7f093197b700 17:45:29.170 [opensc-tool] reader-pcsc.c:612:pcsc_lock: called
0x7f093197b700 17:45:29.170 [opensc-tool] card.c:449:sc_lock: returning with: 0 (Success)
0x7f093197b700 17:45:29.170 [opensc-tool] apdu.c:521:sc_transmit: called
0x7f093197b700 17:45:29.170 [opensc-tool] apdu.c:371:sc_single_transmit: called
0x7f093197b700 17:45:29.170 [opensc-tool] apdu.c:378:sc_single_transmit: CLA:0, INS:A4, P1:0, P2:0, data(2) 0x7ffeb50730f0
0x7f093197b700 17:45:29.170 [opensc-tool] reader-pcsc.c:283:pcsc_transmit: reader 'MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00'
0x7f093197b700 17:45:29.170 [opensc-tool] reader-pcsc.c:284:pcsc_transmit: 
Outgoing APDU (7 bytes):
00 A4 00 00 02 3F 00 .....?.
0x7f093197b700 17:45:29.170 [opensc-tool] reader-pcsc.c:212:pcsc_internal_transmit: called
0x7f093197b700 17:45:29.425 [opensc-tool] reader-pcsc.c:293:pcsc_transmit: 
Incoming APDU (2 bytes):
6A 86 j.
0x7f093197b700 17:45:29.425 [opensc-tool] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7f093197b700 17:45:29.425 [opensc-tool] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7f093197b700 17:45:29.425 [opensc-tool] card.c:459:sc_unlock: called
0x7f093197b700 17:45:29.425 [opensc-tool] reader-pcsc.c:662:pcsc_unlock: called
0x7f093197b700 17:45:29.428 [opensc-tool] iso7816.c:124:iso7816_check_sw: Incorrect parameters P1-P2
0x7f093197b700 17:45:29.428 [opensc-tool] card-iasecc.c:270:iasecc_select_mf: Cannot select MF: -1205 (Incorrect parameters in APDU)
0x7f093197b700 17:45:29.428 [opensc-tool] card-iasecc.c:882:iasecc_select_file: MF selection error: -1205 (Incorrect parameters in APDU)
0x7f093197b700 17:45:29.428 [opensc-tool] card.c:770:sc_select_file: 'SELECT' error: -1205 (Incorrect parameters in APDU)
0x7f093197b700 17:45:29.428 [opensc-tool] ef-atr.c:151:sc_parse_ef_atr: Cannot select EF(ATR) file: -1205 (Incorrect parameters in APDU)
0x7f093197b700 17:45:29.428 [opensc-tool] card-iasecc.c:367:iasecc_parse_ef_atr: MF selection error: -1205 (Incorrect parameters in APDU)
0x7f093197b700 17:45:29.428 [opensc-tool] card-iasecc.c:541:iasecc_init_sagem: IASECC: ATR parse failed: -1205 (Incorrect parameters in APDU)
0x7f093197b700 17:45:29.428 [opensc-tool] card-iasecc.c:678:iasecc_init: returning with: -1205 (Incorrect parameters in APDU)
0x7f093197b700 17:45:29.428 [opensc-tool] card.c:262:sc_connect_card: driver 'IAS-ECC' init() failed: Incorrect parameters in APDU
0x7f093197b700 17:45:29.950 [opensc-tool] reader-pcsc.c:597:pcsc_disconnect: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00:SCardDisconnect returned: 0x00000000
0x7f093197b700 17:45:29.950 [opensc-tool] card.c:335:sc_connect_card: returning with: -1205 (Incorrect parameters in APDU)
0x7f093197b700 17:45:29.950 [opensc-tool] ctx.c:870:sc_release_context: called
0x7f093197b700 17:45:29.950 [opensc-tool] reader-pcsc.c:896:pcsc_finish: called
```

**Comment by achelouah** [comment]
... and my `opensc.conf`

```
[root@localhost ~]# cat /etc/opensc.conf
# Configuration file for OpenSC
# Example configuration file

# NOTE: All key-value pairs must be terminated by a semicolon.

# Default values for any application
# These can be overridden by an application
# specific configuration block.
app default {
        # Amount of debug info to print
        #
        # A greater value means more debug info.
        # Default: 0
        #
        debug = 7;

        # The file to which debug output will be written
        #
        # Special values 'stdout' and 'stderr' are recognized.
        # Default: stderr
        #
        # debug_file = /tmp/opensc-debug.log
        debug_file = /tmp/opensc-debug.log

        # Reopen debug file at the every debug message.
        #
        # For Windows it's forced to 'true'. The reason is that
        #   in Windows file handles can not be shared between DLL-s,
        #   each DLL has a separate file handle table.
        #
        # Default: false
        # reopen_debug_file = true;

        # PKCS#15 initialization / personalization
        # profiles directory for pkcs15-init.
        # Default: /opt/opensc/share/opensc
        #
        # profile_dir = /opt/opensc/share/opensc;

        # Paranoid memory allocation.
        #
        # If set to 'true', then refuse to continue when locking of non-pageable
        # memory fails. This can cause subtle failures but is more secure when
        # you have a swap disk.
        # Default: false
        #
        # paranoid_memory = false;

        # Dsiable pop-ups of built-in GUI
        #
        # Default: false
        # disable_popups = true;

        # Enable default card driver
        # Default card driver is explicitely enabled for the 'opensc-explorer' and 'opensc-tool'.
        #
        # Default: false
        # enable_default_driver = true;

        # CT-API module configuration.
        reader_driver ctapi {
                # module /opt/opensc/lib/libtowitoko.so {
                        # CT-API ports:
                        # 0..3          COM1..4
                        # 4             Printer
                        # 5             Modem
                        # 6..7          LPT1..2
                        # ports = 0;
                # }
        }

        # The following section shows definitions for PC/SC readers.
        reader_driver pcsc {
                # Limit command and response sizes. Some Readers don't propagate their
                # transceive capabilities correctly. max_send_size and max_recv_size
                # allow setting the limits manually, for example to enable extended
                # length capabilities.
                # Default: max_send_size = 255, max_recv_size = 256;
                # max_send_size = 65535;
                # max_recv_size = 65536;
                #
                # Connect to reader in exclusive mode?
                # Default: false
                # connect_exclusive = true;
                #
                # What to do when disconnecting from a card (SCardDisconnect)
                # Valid values: leave, reset, unpower.
                # Default: reset
                # disconnect_action = unpower;
                #
                # What to do at the end of a transaction (SCardEndTransaction)
                # Valid values: leave, reset, unpower.
                # Default: leave
                # transaction_end_action = reset;
                #
                # What to do when reconnection to a card (SCardReconnect)
                # Valid values: leave, reset, unpower.
                # Note that this affects only the internal reconnect (after a SCARD_W_RESET_CARD).
                # A forced reset via sc_reset() always does a full powerup.
                # Default: leave
                # reconnect_action = reset;
                #
                # Enable pinpad if detected (PC/SC v2.0.2 Part 10)
                # Default: true
                # enable_pinpad = false;
                #
                # Detect reader capabilities with escape commands (wrapped APDUs with
                # CLA=0xFF as defined by PC/SC pt. 3 and BSI TR-03119, e.g. for getting
                # the UID, escaped PIN commands and the reader's firmware version)
                # Default: false
                # enable_escape = true;
                #
                # Use specific pcsc provider.
                # Default: libpcsclite.so.1
                # provider_library = libpcsclite.so.1
        }

        # Options for OpenCT support
        reader_driver openct {
                # Virtual readers to allocate.
                # Default: 2
                # readers = 5;
                #
                # Limit command and response sizes.
                # Default: n/a
                # max_send_size = 255;
                # max_recv_size = 256;
        }

        # Options for CryptoTokenKit support
        reader_driver cryptotokenkit {
                # Limit command and response sizes. Some Readers don't propagate their
                # transceive capabilities correctly. max_send_size and max_recv_size
                # allow setting the limits manually, for example to enable extended
                # length capabilities.
                # Default: autodetect
                # max_send_size = 65535;
                # max_recv_size = 65536;
        }

        # Whitelist of card drivers to load at start-up
        #
        # The supported internal card driver names can be retrieved
        # from the output of:
        # $ opensc-tool --list-drivers
        #
        # A special value of 'internal' will load all
        # statically linked drivers. If an unknown (ie. not
        # internal) driver is supplied, a separate configuration
        # configuration block has to be written for the driver.
        # Default: internal
        # NOTE: When "internal" keyword is used, must be last entry
        #
        card_drivers = npa, internal;

        # Card driver configuration blocks.

        # For card drivers loaded from an external shared library/DLL,
        # you need to specify the path name of the module
        #
        # card_driver customcos {
                # The location of the driver library
                # module = /opt/opensc/lib/libcard_customcos.so;
        # }

        card_driver npa {
                # German ID card requires the CAN to be verified before QES PIN. This,
                # however, is not part of the PKCS#15 profile of the card. So for
                # verifying the QES PIN we actually need both. The CAN may be given
                # here. If the CAN is not given here, it will be prompted on the
                # command line or on the reader (depending on the reader's
                # capabilities).
                #
                #can = 222222;

                # QES is only possible with a Comfort Reader (CAT-K), which holds a
                # cryptographic key to authenticate itself as signature terminal (ST).
                # We usually will use the reader's capability to sign the data.
                # However, during developement you may specify soft certificates and
                # keys for a ST below.
                # The following example EAC PKI can be found in vicc's example data:
                # https://github.com/frankmorgner/vsmartcard/tree/master/virtualsmartcard/npa-example-data
                #
                #st_dv_certificate = ZZSTDVCA00001.cvcert;
                #st_certificate = ZZSTTERM00001.cvcert;
                #st_key = ZZSTTERM00001.pkcs8;
        }

        # Force using specific card driver
        #
        # If this option is present, OpenSC will use the supplied
        # driver with all inserted cards.
        #
        # Default: autodetect
        #
        # force_card_driver = customcos;

        # Configuration block for DNIe
        #
        # Card DNIe has an option to show an extra warning before
        # issuing a signature.

        card_driver dnie {
                # Disable / enable warning message when performing a
                # signature operation with the DNIe.
                # Only used if compiled with --enable-dnie-ui
                # user_consent_enabled = yes;

                # Specify the pinentry application to use if warning
                # is configured to be displayed using pinentry.
                # Default: /usr/bin/pinentry
                # Only used if compiled with --enable-dnie-ui
                # user_consent_app = "/usr/bin/pinentry";
        }

        # In addition to the built-in list of known cards in the
        # card driver, you can configure a new card for the driver
        # using the card_atr block. The goal is to centralize
        # everything related to a certain card to card_atr.
        #
        # The supported internal card driver names can be retrieved
        # from the output of:
        # $ opensc-tool --list-drivers

        # Generic format: card_atr <hex encoded ATR (case-sensitive!)>

        # New card entry for the flex card driver
        # card_atr 3b:f0:0d:ca:fe {
                # All parameters for the context are
                # optional unless specified otherwise.

                # Context: global, card driver
                #
                # ATR mask value
                #
                # The mask is logically AND'd with an
                # card ATR prior to comparison with the
                # ATR reference value above. Using mask
                # allows identifying and configuring
                # multiple ATRs as the same card model.
                # atrmask = "ff:ff:ff:ff:ff";

                # Context: card driver
                #
                # Specify used card driver (REQUIRED).
                #
                # When enabled, overrides all possible
                # settings from the card drivers built-in
                # card configuration list.
                # driver = "flex";

                # Set card name for card drivers that allows it.
                # name = "My CryptoFlex card";

                # Card type as an integer value.
                #
                # Depending on card driver, this allows
                # tuning the behaviour of the card driver
                # for your card.
                # type = "2002";

                # Card flags as an hex value.
                # Multiple values are OR'd together.
                #
                # Depending on card driver, this allows
                # fine-tuning the capabilities in
                # the card driver for your card.
                #
                # Optionally, some known parameters
                # can be specified as strings:
                #
                # rng - On-board random number source
                #
                # flags = "rng", "0x80000000";

                # Enable pkcs11 initialization.
                # Default: no
                # pkcs11_enable_InitToken = yes;

                #
                # Context: PKCS#15 emulation layer
                #
                # When using PKCS#15 emulation, force
                # the emulation driver for specific cards.
                #
                # Required for external drivers, but can
                # be used with built-in drivers, too.
                # pkcs15emu = "custom";

                #
                # Context: reader driver
                #
                # Force protocol selection for specific cards.
                # Known parameters: t0, t1, raw
                # force_protocol = "t0";

                # Context: minidriver
                #
                # md_read_only: Mark card as read/only card in Minidriver/BaseCSP interface (Default: false)
                # md_supports_X509_enrollment: Indicate X509 enrollment support at Minidriver/BaseCSP interface (Default: false)
                # md_guid_as_id: Use the GUID generated for the key as id in the PKCS#15 structure (Default: false - auto generated)
                # md_guid_as_label: Use the GUID generated for the key as label in the PKCS#15 structure (Default: false - no label set)
                # md_supports_container_key_gen: Card allows generating key pairs on the card (Default: false)
                # md_supports_container_key_import: Card allows importing private keys (Default: false)
        # }

        # PIV cards need an entry similar to this one:
        # card_atr 3B:7D:96:00:00:80:31:80:65:B0:83:11:00:AC:83:00:90:00 {
                # name = "PIV-II";
                # driver = "piv";
        # }

        # Estonian ID card and Micardo driver sometimes only play together with T=0
        # In theory only the 'cold' ATR should be specified, as T=0 will
        # be the preferred protocol once you boot it up with T=0, but be
        # paranoid.
        #
        # Warm ATR v1
        card_atr 3b:6e:00:ff:45:73:74:45:49:44:20:76:65:72:20:31:2e:30 {
                force_protocol = t0;
        }
        # Cold ATR v1
        card_atr 3b:fe:94:00:ff:80:b1:fa:45:1f:03:45:73:74:45:49:44:20:76:65:72:20:31:2e:30:43 {
                force_protocol = t0;
        }
        # Warm ATR v2
        card_atr 3b:5e:11:ff:45:73:74:45:49:44:20:76:65:72:20:31:2e:30 {
                force_protocol = t0;
        }
        # Cold ATR v2
        card_atr 3b:de:18:ff:c0:80:b1:fe:45:1f:03:45:73:74:45:49:44:20:76:65:72:20:31:2e:30:2b {
                force_protocol = t0;
        }
        # Digi-ID cold ATR. The same card has the same warm ATR as "Cold ATR v1" above
        # The card is claimed to only support T=0 but in fact (sometimes) works with T=1, even if not advertised in ATR.
        card_atr 3b:6e:00:00:45:73:74:45:49:44:20:76:65:72:20:31:2e:30 {
                force_protocol = t0;
        }

        # D-Trust cards are also based on micardo and need T=0 for some reason
        card_atr 3b:ff:94:00:ff:80:b1:fe:45:1f:03:00:68:d2:76:00:00:28:ff:05:1e:31:80:00:90:00:23 {
                force_protocol = t0;
        }
        card_atr 3b:ff:11:00:ff:80:b1:fe:45:1f:03:00:68:d2:76:00:00:28:ff:05:1e:31:80:00:90:00:a6 {
                force_protocol = t0;
        }

        # Oberthur's AuthentIC v3.2.2
        card_atr 3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:00:70:0A:90:00:8B {
                type = 11100;
                driver = "authentic";
                name = "AuthentIC v3.1";

                # Name of SM configuration sub-section
                # secure_messaging = local_authentic;
        }

        # IAS/ECC cards
        card_atr 3B:7F:96:00:00:00:31:B9:64:40:70:14:10:73:94:01:80:82:90:00 {
                type = 25001;
                driver = "iasecc";
                name = "Gemalto MultiApp IAS/ECC v1.0.1";
                secure_messaging = local_gemalto_iam;
                # secure_messaging = local_adele;
                md_read_only = false;
                md_supports_X509_enrollment = true;
        }
        card_atr 3B:7F:96:00:00:00:31:B8:64:40:70:14:10:73:94:01:80:82:90:00 {
                type = 25001;
                driver = "iasecc";
                name = "Gemalto MultiApp IAS/ECC v1.0.1";
                secure_messaging = local_gemalto_iam;
                md_read_only = false;
                md_supports_X509_enrollment = true;
        }
        #card_atr 3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:08:00:07:90:00:FE {
        #       type = 25002;
        #       driver = "iasecc";
        #       name = "Oberthur IAS/ECC v1.0.1";
        #       # No 'admin' application for this card -- no secure messaging
        #}
        #card_atr 3B:7F:18:00:00:00:31:B8:64:50:23:EC:C1:73:94:01:80:82:90:00 {
        #       type = 25003;
        #       driver = "iasecc";
        #       name = "Morpho YpsID S3 IAS/ECC";
        #       # secure_messaging = local_morpho_YpsID_S3;
        #}
        card_atr 3B:7D:18:00:02:80:57:59:50:53:49:44:30:34:83:7F:90:00 {
                type = 25003;
                driver = "iasecc";
                name = "Morpho ypsID Key E IAS/ECC";
                # secure_messaging = local_morpho_YpsID_S3;
        }
        #card_atr 3B:DF:96:00:80:31:FE:45:00:31:B8:64:04:1F:EC:C1:73:94:01:80:82:90:00:EC {
        #       type = 25005;
        #       driver = "iasecc";
        #       name = "Morpho MI IAS/ECC v1.0.1";
        #       md_read_only = false;
        #       md_supports_X509_enrollment = true;
        #       secure_messaging = local_morpho_mi;
        #}
        card_atr 3B:DF:18:FF:81:91:FE:1F:C3:00:31:B8:64:0C:01:EC:C1:73:94:01:80:82:90:00:B3 {
                type = 25004;
                driver = "iasecc";
                name = "Amos IAS/ECC v1.0.1";
                md_read_only = false;
                md_supports_X509_enrollment = true;
                secure_messaging = local_amos;
        }
        card_atr 3B:DC:18:FF:81:91:FE:1F:C3:80:73:C8:21:13:66:01:0B:03:52:00:05:38 {
                type = 25004;
                driver = "iasecc";
                name = "Amos IAS/ECC v1.0.1";
                md_read_only = false;
                md_supports_X509_enrollment = true;
                secure_messaging = local_amos_eid;
        }

        # SmartCard-HSM with contact-based interface or USB-Stick
        card_atr 3B:FE:18:00:00:81:31:FE:45:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:FA {
                driver = "sc-hsm";
                md_read_only = false;
                md_supports_X509_enrollment = true;
                md_supports_container_key_gen = true;
                md_guid_as_label = true;
        }

        # SmartCard-HSM with contact-less interface
        card_atr 3B:8E:80:01:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:18 {
                driver = "sc-hsm";
                md_read_only = false;
                md_supports_X509_enrollment = true;
                md_supports_container_key_gen = true;
                md_guid_as_label = true;
        }

        secure_messaging local_authentic  {
                # name of external SM module
                # module_name = libsmm-local.so;
                # directory with external SM module
                # Default: defined by windows register
                module_path = /opt/opensc/lib;

                # specific data to tune the module initialization
                # module_data = "Here can be your SM module init data";

                # SM mode:
                # 'transmit' -- in this mode the procedure to securize an APDU is called by the OpenSC general
                #              APDU transmit procedure.
                #               In this mode all APDUs, except the ones filtered by the card specific procedure,
                #              are securized.
                # 'acl' -- in this mode APDU are securized only if needed by the ACLs of the command to be executed.
                #
                #mode = transmit;

                # SM type specific flags
                # flags = 0x78;    # 0x78 -- level 3, channel 0

                # Default KMC of the GP Card Manager for the Oberthur's Java cards
                # kmc = "00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00";
        }

        secure_messaging local_gemalto_iam  {
                module_name = libsmm-local.so;
                module_path = /opt/opensc/lib;
                # module_data = "";
                type = acl;          # transmit, acl

                ifd_serial = "11:22:33:44:55:66:77:88";

                # Keyset values from IAM profiles of the Gemalto IAS/ECC cards
                keyset_02_enc = "RW_PRIV_ENC_TEST";
                keyset_02_mac = "RW_PRIV_MAC_TEST";

                keyset_E828BD080FD2504543432D654944_01_enc = "RO_ENC_TEST_KEY_";
                keyset_E828BD080FD2504543432D654944_01_mac = "RO_MAC_TEST_KEY_";

                keyset_E828BD080FD2504543432D654944_03_enc = "RW_PUBL_ENC_TEST";
                keyset_E828BD080FD2504543432D654944_03_mac = "RW_PUBL_MAC_TEST";
        }

        secure_messaging local_amos  {
                module_name = libsmm-local.so;
                module_path = /opt/opensc/lib;
                # module_data = "";
                mode = acl;
                ifd_serial = "11:22:33:44:55:66:77:88";
                keyset_02_enc = "ENCROECHANTILLON";
                keyset_02_mac = "MACROECHANTILLON";
        }

        secure_messaging local_amos_eid  {
                module_name = libsmm-local.so;
                module_path = /opt/opensc/lib;
                # module_data = "";
                mode = acl;
                ifd_serial = "11:22:33:44:55:66:77:88";
                keyset_E828BD080FD2504543432D654944_03_enc = "RW_PUBL_ENC_TEST";
                keyset_E828BD080FD2504543432D654944_03_mac = "RW_PUBL_MAC_TEST";
        }

        secure_messaging local_adele  {
                module_name = libsmm-local.so;
                module_path = /opt/opensc/lib;
                # module_data = "";
                type = acl;          # transmit, acl

                ifd_serial = "11:22:33:44:55:66:77:88";

                # Keyset values from 'Adele' profiles of the IAS/ECC cards
                keyset_01_enc = "EMENCECHANTILLON";
                keyset_01_mac = "EMMACECHANTILLON";

                keyset_02_enc = "AAENCECHANTILLON";
                keyset_02_mac = "AAMACECHANTILLON";

                keyset_E828BD080FD2500000040301_02_enc = "E2ENCECHANTILLON";
                keyset_E828BD080FD2500000040301_02_mac = "E2MACECHANTILLON";

                keyset_D2500000044164E86C650101_02_enc = "E1ENCECHANTILLON";
                keyset_D2500000044164E86C650101_02_mac = "E1MACECHANTILLON";

                keyset_D2500000044164E86C650101_03_enc = "SIENCECHANTILLON";
                keyset_D2500000044164E86C650101_03_mac = "SIMACECHANTILLON";
        }

        # Below are the framework specific configuration blocks.

        # PKCS #15
        framework pkcs15 {
                # Whether to use the cache files in the user's
                # home directory.
                #
                # At the moment you have to 'teach' the card
                # to the system by running command: pkcs15-tool -L
                #
                # WARNING: Caching shouldn't be used in setuid root
                # applications.
                # Default: false
                # use_file_caching = true;
                #
                # set a path for caching
                # so you do not use the env variables and for pam_pkcs11
                # (with certificate check)  where $HOME is not set
                # Default: path in user home
                # file_cache_dir = /var/lib/opensc/cache
                #
                # Use PIN caching?
                # Default: true
                # use_pin_caching = false;
                #
                # How many times to use a PIN from cache before re-authenticating it?
                # Default: 10
                # pin_cache_counter = 3;
                #
                # Older PKCS#11 applications not supporting CKA_ALWAYS_AUTHENTICATE
                # may need to set this to get signatures to work with some cards.
                # Default: false
                # pin_cache_ignore_user_consent = true;
                #
                # Enable pkcs15 emulation.
                # Default: yes
                # enable_pkcs15_emulation = no;
                #
                # Prefer pkcs15 emulation code before
                # the normal pkcs15 processing.
                # Some cards (like esteid and pteid) work in emu-only mode,
                # and do not depend on this option.
                #
                # Default: no
                # try_emulation_first = yes;

                # Enable builtin emulators.
                # Default: yes
                # enable_builtin_emulation = no;
                #
                # List of the builtin pkcs15 emulators to test
                # Default: esteid, openpgp, tcos, starcert, itacns, infocamere, postecert, actalis, atrust-acos, gemsafeGPK, gemsafeV1, tccardos, PIV-II;
                # builtin_emulators = openpgp;

                # additional settings per driver
                #
                # For pkcs15 emulators loaded from an external shared
                # library/DLL, you need to specify the path name of the module
                # and customize the card_atr example above correctly.
                #
                # emulate custom {
                        # The location of the driver library
                        # module = /opt/opensc/lib/libp15emu_custom.so;
                # }

                # some additional application parameters:
                # - type (generic, protected) used to distinguish the common access application
                #   and application for which authentication to perform some operation cannot be
                #   obtained with the common procedures (ex. object creation protected by secure messaging).
                #   Used by PKCS#11 module configurated to expose restricted number of slots.
                #   (for ex. configurated to expose only User PIN slot, User and Sign PINs slots, ...)
                #
                # - disable: do not expose application in PKCS15 framework
                #            default 'false'
                application E828BD080FD25047656E65726963 {
                        type = generic;
                        model = "ECC Generic PKI";
                        # disable = true
                }

                application E828BD080FD2500000040301 {
                        type = generic;
                        model = "Adèle Générique";
                }

                application E828BD080FD2504543432D654944 {
                        type = protected;
                        model = "ECC eID";
                }

                application E828BD080FD2500000040201 {
                        type = protected;
                        model = "Adèle Admin-2";
                }
        }
}

# Parameters for the OpenSC PKCS11 module
app opensc-pkcs11 {
        pkcs11 {
                # Maximum Number of virtual slots.
                # If there are more slots than defined here,
                # the remaining slots will be hidden from PKCS#11.
                # Default: 16
                # max_virtual_slots = 32;

                # Maximum number of slots per smart card.
                # If the card has fewer keys than defined here,
                # the remaining number of slots will be empty.
                # Default: 4
                # slots_per_card = 2;

                # (max_virtual_slots/slots_per_card) limits the number of readers
                # that can be used on the system. Default is then 16/4=4 readers.

                # Normally, the pkcs11 module will create
                # the full number of slots defined above by
                # num_slots. If there are fewer pins/keys on
                # the card, the remaining keys will be empty
                # (and you will be able to create new objects
                # within them).
                # Default: true
                # hide_empty_tokens = false;

                # By default, the OpenSC PKCS#11 module will not lock your card
                # once you authenticate to the card via C_Login.
                #
                # Thus the other users or other applications is not prevented
                # from connecting to the card and perform crypto operations
                # (which may be possible because you have already authenticated
                # with the card). This setting is not very secure.
                #
                # Also, if your card is not locked, you can enconter problems
                # due to limitation of the OpenSC framework, that still is not
                # thoroughly tested in the multi threads environment.
                #
                # Your settings will be more secure if you choose to lock your
                # card. Nevertheless this behavior is a known violation of PKCS#11
                # specification. Now once one application has started using your
                # card with C_Login, no other application can use it, until
                # the first is done and calls C_Logout or C_Finalize. In the case
                # of many PKCS#11 application this does not happen until you exit
                # the application.
                # Thus it is impossible to use several smart card aware applications
                # at the same time, e.g. you cannot run both Firefox and Thunderbird at
                # the same time, if both are configured to use your smart card.
                #
                # Default: false
                # lock_login = true;

                # By default, interacting with the OpenSC PKCS#11 module may change the
                # state of the token, e.g. whether a user is logged in or not.
                #
                # Thus other users or other applications may change or use the state of
                # the token unknowingly. Other applications may create signatures
                # abusing an existing login or they may logout unnoticed.
                #
                # With this setting enabled the login state of the token is tracked and
                # cached (including the PIN). Every transaction is preceeded by
                # restoring the login state.  After every transaction a logout is
                # performed. This setting by default also enables `lock_login` (see
                # above) to disable access for other applications during the atomic
                # transactions.
                #
                # Please note that any PIN-pad should be disabled (see `enable_pinpad`
                # above), because the user would have to input his PIN for every
                # transaction.
                #
                # Default: false
                # atomic = true;

                # With this setting disabled, the OpenSC PKCS#11 module will initialize
                # the slots available when the application calls `C_GetSlotList`. With
                # this setting enabled, the slots will also get initialized when
                # C_GetSlotInfo is called.
                #
                # This setting is a workaround for Java which does not call
                # `C_GetSlotList` when configured with a static `slot` instead of
                # `slotListIndex`.
                #
                # Default: true
                # init_sloppy = false;

                # User PIN unblock style
                #    none:  PIN unblock is not possible with PKCS#11 API;
                #    set_pin_in_unlogged_session:  C_SetPIN() in unlogged session:
                #       PUK is passed as the 'OldPin' argument of the C_SetPIN() call.
                #    set_pin_in_specific_context:  C_SetPIN() in the CKU_SPECIFIC_CONTEXT logged session:
                #       PUK is passed as the 'OldPin' argument of the C_SetPIN() call.
                #    init_pin_in_so_session:  C_InitPIN() in CKU_SO logged session:
                #       User PIN 'UNBLOCK' is protected by SOPIN. (PUK == SOPIN).
                #       # Actually this style works only for the PKCS15 contents without SOPIN.
                #       # For those with SOPIN, this mode will be usefull for the cards without
                #       #   modes 00 and 01 of ISO command 'RESET RETRY COUNTER'. --vt
                #
                # Default: none
                # user_pin_unblock_style = set_pin_in_unlogged_session;

                # Create slot for unblocking PIN with PUK
                # This way PKCS#11 API can be used to login with PUK and
                # change a PIN.
                # Warning: causes problems with some applications like
                # firefox and thunderbird. Thus turned off by default
                #
                # Default: false
                # create_puk_slot = true;

                # Report as 'zero' the CKA_ID attribute of CA certificate
                # For the unknown reason the middleware of the manufacturer of gemalto (axalto, gemplus)
                # card reports as '0' the CKA_ID of CA cartificates.
                # Maybe someone else will need it. (Would be nice to know who and what for -- VTA)
                #
                # Default: false
                # zero_ckaid_for_ca_certs = true;

                # List of readers to ignore
                # If any of the strings listed below is matched (case sensitive) in a reader name,
                # the reader is ignored by the PKCS#11 module.
                #
                # Default: empty
                # ignored_readers = "CardMan 1021", "SPR 532";

                # Symbolic names of PINs for which slots are created
                # Card can contain more then one PINs or more then one on-card application with
                #   its own PINs. Normally, to access all of them with the PKCS#11 API a slot has to be
                #   created for all of them. Many slots could be ennoying for some of widely used application,
                #   like FireFox. This configuration parameter allows to select the PIN(s)
                #   for which PKCS#11 slot will be created.
                # Actually recognised following symbolic names:
                #  'user', 'sign', 'all'
                # Only PINs initialised, non-SoPIN, non-unblocking are associated with symbolic name.
                # 'user' is identified as first global or first local PIN.
                # 'sign' is identified as second PIN: first local, second global or second local.
                # 'all' slot created for all non-sopin, non-unblocking PINs,
                #   optionally for PUK (see option 'create_puk_slot')
                #
                # Default: all
                # create_slots_for_pins = "user,sign";
                # create_slots_for_pins = "sign";
                #
                # For the module to simulate the opensc-onepin module behavior the following option
                # must be set:
                # create_slots_for_pins = "user"
        }
}

app onepin-opensc-pkcs11 {
        pkcs11 {
                slots_per_card = 1;
        }
}

# Used by OpenSC.tokend on Mac OS X only
app tokend {
        # The file to which debug log will be written
        # Default: /tmp/opensc-tokend.log
        #
        # debug_file = /Library/Logs/OpenSC.tokend.log

        framework tokend {
                # Score for OpenSC.tokend
                # The tokend with the highest score shall be used.
                # Default: 300
                #
                # score = 10;
        }
}

# Used by OpenSC minidriver on Windows only
app cardmod {
}
```

**Comment by frankmorgner** [comment]
Try connecting with t=1

**Comment by achelouah** [comment]
Thank you for your help. But can you be more precise about `connecting with t=1`. You mean `type=1` in `opensc.conf` ?

```
        card_atr 3B:7D:18:00:02:80:57:59:50:53:49:44:30:34:83:7F:90:00 {
                type = 1;
                driver = "iasecc";
                name = "Morpho ypsID Key E IAS/ECC";
                # secure_messaging = local_morpho_YpsID_S3;
        }
```

**Comment by frankmorgner** [comment]
use force_protocol in the atr block:
```
        # Force protocol selection for specific cards.
        # Known parameters: t0, t1, raw
        force_protocol = "t1";
```

**Comment by achelouah** [comment]
I tried with

```
        card_atr 3B:7D:18:00:02:80:57:59:50:53:49:44:30:34:83:7F:90:00 {
                force_protocol = t1;
        }
```
in `opensc.conf`.

```
[root@localhost ~]# opensc-tool -l -n
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    Yes             MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00
Using reader with a card: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00
Failed to connect to card: Generic reader error
```

See the attached file.
[opensc-debug.txt](https://github.com/OpenSC/OpenSC/files/1138803/opensc-debug.txt)

**Comment by frankmorgner** [comment]
When connecting with T=1 instead of T=0, PC/SC returns `SCARD_E_PROTO_MISMATCH` (`0x8010000F`): "The requested protocols are incompatible with the protocol currently in use with the smart card."

Did you forget to mention some concurrent process that holds a handle to the token with T=0?

If not, your token seems to only support T=0. This is strange, because your card does not seem to support T=0-like communication. You've previously reported this error:
```
Outgoing APDU (7 bytes):
00 A4 00 00 02 3F 00 .....?.
0x7f093197b700 17:45:29.170 [opensc-tool] reader-pcsc.c:212:pcsc_internal_transmit: called
0x7f093197b700 17:45:29.425 [opensc-tool] reader-pcsc.c:293:pcsc_transmit: 
Incoming APDU (2 bytes):
6A 86 j.
0x7f093197b700 17:45:29.425 [opensc-tool] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7f093197b700 17:45:29.425 [opensc-tool] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7f093197b700 17:45:29.425 [opensc-tool] card.c:459:sc_unlock: called
0x7f093197b700 17:45:29.425 [opensc-tool] reader-pcsc.c:662:pcsc_unlock: called
0x7f093197b700 17:45:29.428 [opensc-tool] iso7816.c:124:iso7816_check_sw: Incorrect parameters P1-P2
0x7f093197b700 17:45:29.428 [opensc-tool] card-iasecc.c:270:iasecc_select_mf: Cannot select MF: -1205 (Incorrect parameters in APDU)
0x7f093197b700 17:45:29.428 [opensc-tool] card-iasecc.c:882:iasecc_select_file: MF selection error: -1205 (Incorrect parameters in APDU)
```
We could go on in discussing tiny modifications of OpenSC's behavior, but I think you need to find someone to have a deeper look, here.

**Comment by achelouah** [comment]
There is no concurrent process that possibly holds a handle to the token. Actually, I tested the token on two different computers with the same result. With `T=0`, I get the message

```
[root@localhost ~]# opensc-tool -l -n
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    Yes             MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00
Using reader with a card: MORPHO TOKEN EM (1305400000317523-23 030C 5426 2001 4154) 00 00
Unsupported card
```
See attached file


[opensc-debug.txt](https://github.com/OpenSC/OpenSC/files/1140205/opensc-debug.txt)

**Comment by frankmorgner** [comment]
Sorry, if no IAS/ECC developer is available, then, unfortunately, we can't do anything for you.

**Comment by achelouah** [comment]
Thanks again for your support.

**Comment by frankmorgner** [comment]
Seems your token is simply unsupported...
